---
title: "Starter Notebook"
author: "You, Scientist"
format: html
execute:
  keep-md: true
---

```{r}
#Load in any packages you need
library(tidyverse)

#Read in any data

```

```{r}
instructors <- c("Duryea", "Gilbert")

instructors
```

Challenge 1: List of nucleotides

```{r}
nucleotides <- c("A", "G","C","T")
nucleotides
```

The code block combines the strings of the four nucleotides into the variable nucleotides and then prints it out.

Challenge 2: Random string of 15 nucleotides

```{r}
numNucleotide <- 15

randGenome <- sample(nucleotides, size = numNucleotide, replace = TRUE)

randGenome
```

A length of a genome is set at 15 with the variable numNucleotide. The variable randGenome is sampled (randomized) with the size of 15. Randgenome is printed out.

Challenge 3

```{r}
numNucleotide <- 1500

randGenome <- sample(nucleotides, size = numNucleotide, replace = TRUE)

randGenome <- paste(randGenome, collapse ="")

randGenome

```

The length of nucleotides is set at 1500. The only difference between this and challenge 2 is that string is collapsed to display a sequence of letters instead of the full names (adenosine, guanine, thymine, cytosine).

Challenge 4: Generate a random genome of 100 nucleotides and count adenosine

```{r}
set.seed(215)
numNucleotide <- 100

randGenome <- sample(nucleotides, size = numNucleotide, replace = TRUE)

randGenome <- paste(randGenome, collapse ="")

randGenome
```

Challenge 5:

```{r}
myProduct <- 1

for(j in 1:15){
  myProduct <- myProduct + j
  print(myProduct)
}


```

Creates a for loop where it runs through the variable j 15 times and whatever number it produces is added to myProduct

## Challenge 6

```{r}
#Use this code chunk to complete Challenge 6
genomeLength <- 20
randGenome <- sample(nucleotides, size = genomeLength, replace = TRUE)
randGenome <- paste(randGenome, collapse = "")
randGenome
  for(j in 1:10){
  print(str_sub(randGenome, start = j, end = j))
}

```

Setting the genome length to 20, we create a randomized sample based on the length and collapse the string to the letters A,G,C,T. We create a for loop for randGenome that is printed 10 times.

## Challenge 7

```{r}
randGenome
Adenine_count <- 0
for (i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A"){
    Adenine_count <- Adenine_count + 1
    print(Adenine_count)
}
}
```

Using the for loop from challenge six, we have the code look for the string "A" or adenine in the genome and add it to a variable called Adenine count. We then print out the final adenine count.

```{r}
randGenome
Adenine_count <- 0
Thymine_count <- 0
Guanine_count <- 0
Cytosine_count <- 0
for (i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A"){
    Adenine_count <- Adenine_count + 1
}
  if(str_sub(randGenome,start = i, end = i) == "G"){
      Guanine_count <- Guanine_count + 1
  }
  if(str_sub(randGenome, start = i, end = i) == "C"){
    Cytosine_count <- Cytosine_count + 1
  }
  if(str_sub(randGenome, start = i, end = i) == "T"){
    Thymine_count <- Thymine_count+1
}
  print(c(Cytosine_count,Thymine_count,Guanine_count,Adenine_count))
  }      


```

The only options are AGCT. Nothing else is possible. We are already guaranteening we are covering all the conditions. If your can split the data into a dichotomy: else is otherwise instead of restrictive. This expands on challenge 7 but adds in a frequency for the other nucleotides as well. The final print statement combines all the variables together.

## Challenge 9

```{r}
vib_c <- scan("C:/Users/Joshua/OneDrive - SNHU/Desktop/Classes Spring 2023/Bioinformatics/VibrioCholerae.txt", what = "character", sep = NULL)
vib_c
Adenine_count <- 0
Thymine_count <- 0
Guanine_count <- 0
Cytosine_count <- 0
for (i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "A"){
    Adenine_count <- Adenine_count + 1
}
  if(str_sub(vib_c,start = i, end = i) == "G"){
      Guanine_count <- Guanine_count + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "C"){
    Cytosine_count <- Cytosine_count + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "T"){
    Thymine_count <- Thymine_count+1
}
} 
print(c(Cytosine_count,Thymine_count,Guanine_count,Adenine_count))

```

Using the code from challenge 8, we import the genome of a cholera bacteria and produce an output that returns the frequency of the nucleotides.

## Challenge 10

```{r}
rosalind <- "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC"
rosalind
Adenine_count <- 0
Thymine_count <- 0
Guanine_count <- 0
Cytosine_count <- 0
for (i in 1:nchar(rosalind)){
  if(str_sub(rosalind, start = i, end = i) == "A"){
    Adenine_count <- Adenine_count + 1
}
  if(str_sub(rosalind,start = i, end = i) == "G"){
      Guanine_count <- Guanine_count + 1
  }
  if(str_sub(rosalind, start = i, end = i) == "C"){
    Cytosine_count <- Cytosine_count + 1
  }
  if(str_sub(rosalind, start = i, end = i) == "T"){
    Thymine_count <- Thymine_count+1
}
} 
print(c(Cytosine_count,Thymine_count,Guanine_count,Adenine_count))
```

```{r}
rosalind <- scan("C:/Users/Joshua/OneDrive - SNHU/Desktop/Classes Spring 2023/Bioinformatics/rosalind_dna.txt", what = "character", sep = NULL)
rosalind
Adenine_count <- 0
Thymine_count <- 0
Guanine_count <- 0
Cytosine_count <- 0
for (i in 1:nchar(rosalind)){
  if(str_sub(rosalind, start = i, end = i) == "A"){
    Adenine_count <- Adenine_count + 1
}
  if(str_sub(rosalind,start = i, end = i) == "G"){
      Guanine_count <- Guanine_count + 1
  }
  if(str_sub(rosalind, start = i, end = i) == "C"){
    Cytosine_count <- Cytosine_count + 1
  }
  if(str_sub(rosalind, start = i, end = i) == "T"){
    Thymine_count <- Thymine_count+1
}
} 
print(c(Cytosine_count,Thymine_count,Guanine_count,Adenine_count))
```

# Origin of Replication

## Challenge 1

The goal of this challenge is to generate a random string of 2000 nucleotides and use the function command to count the frequency of cytosine in the genome string.

```{r}
nucleotide_frequency <- function(genomeString, nucleotide = "A"){
  count <- 0
  for(i in 1:nchar(genomeString)){
    if(str_sub(genomeString, start = i, end = i) == nucleotide){
      count <- count + 1
    }
  }
  return(count)
}

nucleotide_frequency("ACTTGCGGGTATCGAG", "G")
```

    nucleotide_frequency <- function(genomeString, nucleotide = "A"){
      count <- 0
      for(i in 1:nchar(genomeString)){
        if(str_sub(genomeString, start = i, end = i) == nucleotide){
          count <- count + 1
        }
      }
      return(count)
    }

    nucleotide_frequency("ACTTGCGGGTATCGAG", "G")

```{r}

genomeLength <- 2000
randGenome <- sample(nucleotides, size = genomeLength, replace = TRUE)

nucleotide_frequency <- function(randGenome, nucleotides = "C"){
  count <- 0
  for(i in 1:nchar(randGenome)){
    if(str_sub(randGenome, start = i, end = i) == nucleotides){
      count <- count + 1
    }
  }
  return(count)
}

randGenome <- paste(randGenome, collapse = "")
nucleotide_frequency(randGenome, "C")
```

My first attempt was to set the variable genomeLength to 2000. From there I created a random sample based on 2000 nucleotides. Then I created a function that counted the amount of cytosines and then returned a count. The problem is that when I returned my function, I only returned the string in nucleotide frequency (ACGT...ACGT) instead of the variable randGenome. Furthermore, I did not condense the list of nucleotides into a number (i.e the count of cytosine) using the paste function.

## Challenge 2

Build a function `rand_genome()` which takes a single parameter `k`, denoting the number of nucleotides in the genome we wish to generate. Your function should return a single genome string of length k.

```{r}
rand_Genome <- function(genomeString, nucleotides){
    nucleotides <- c("A", "G","C","T")
    rand_Genome <- sample(nucleotides, size = genomeString, replace = TRUE)
    
rand_Genome <- paste(rand_Genome, collapse = "")
print(rand_Genome)}

rand_genome <- rand_Genome(2000, nucleotides)
```

My initial logic was to create a sample function for randGenome where genomelength was set as k. Then I would run a function that would determine genomelength by running a for loop that looks for k and then adds it to the count. I would then paste a collapsed version of the text to produce a count of the genomelength. However the output produces just a string of non-randomized text and wont let me collapse the randGenome.

Turns out the for loop was not necessary. We would be defining the size of the genome length through the variable genomeString.

## Challenge 3

```{r}
generate_3_mers <- function(genomeString){
  #list to hold 3-mers
  list_3_mers <- c()
  
  #run throguh genome and append 3-mers to list
  for(i in 1:(nchar(genomeString) - 2)){
    list_3_mers <- list_3_mers%>%
      append(str_sub(genomeString, start = i, end = i + 2))
  }
  #return list of 3-mers
  return(list_3_mers)
    
}

generate_3_mers(rand_genome)
```

For this challenge, I incorporated the ideas of the for loop and function to collect nucleotides in triplets (threemers) in a string of 2000. In the last function, I defined the list of nucleotides as rand_genome which included a set length (determined by genomestring) and nucleotides (which contained the substrings A,C,G,T to represent the nucleotides). For my new function, I input the argument genomestring (which would run whatever length I defined in the last code block).

Then I defined a list as list_3\_mers as c(). This empty vector would collect three nucleotides in sets to be outputted later. I then created a for loop to run through the entirety of the genomestring (I defined the length in the last code block as 2000). The command i in 1:nchar would run through the entire genomestring regardless of its length. The -2 would ensure that there were no cutoffs and each nucleotide set was in a triplet/threemer. Every for loop would save to the vector list_3\_mers (list_3\_mers \<- list_3\_mers%\>%) and through the append function would take the substrings of genomestring (a,c,g,t) and add 2 to whatever substring to form a triplet. With the for loop run, I returned the list of 3mers.

With the function complete, the final step was to run it with the rand_genome. While the genome string gives us a length, it does not include the specific nucleotides. For that reason, I input the argument rand_genome to run through the entire genome.

The main problem I had with this challenge was due to my coding in the last challenge. At the end of the block, I defined rand_genome but forgot to save the command for reuse by writing rand_genome \<- rand_genome(2000, nucleotides). With that simple fix the code gave the desired output.

## Challenge 4

```{r}
generate_k_mers <- function(genomeString, k){
  #list to hold 3-mers
  list_k_mers <- c()
  
  #run throguh genome and append 3-mers to list
  for(i in 1:(nchar(genomeString) - (k-1))){
    list_k_mers <- list_k_mers%>%
      append(str_sub(genomeString, start = i, end = i + (k-1)))
  }
  #return list of 3-mers
  return(list_k_mers)
    
}

my_genome <- rand_Genome(10)
generate_k_mers(my_genome,8)
```

## Challenge 5

```{r}
count_pattern <- function(genomeString, pattern){
  count <- 0
  k <- nchar(pattern)
  #print(k)
  for(i in 1:nchar(genomeString)){
    #print(str_sub(genomeString, start = i, end = i + k -1))
    if(str_sub(genomeString, start = i, end = i+k-1) == pattern){
      count = count + 1
      #print("hi")
    }
  }
  return(count)
}
my_genome <- rand_Genome(30)
my_genome
count_pattern(my_genome, "T")
```

# Origin of Replication II

## Challenge 1

The goal of this code block is to determine the most frequent set of kmers in a random genome. Kmers is defined as a variable set of nucleotides that are combined. For example a kmer could be one nucleotide or 100 hundred. The length of the kmer is defined by the variable k which is contained in the find_frequent_kmers. The other argument in the function is genome which will be defined as a random genome value.

```{r}
find_frequent_kmers <- function(genome,k){
  #Get the Kmers
  kmers <- generate_k_mers(genome,k)
  Kmers <- unique(kmers)
  #count occurences 
  Kmer_counts <- rep(0,length(Kmers))
  for(i in 1:length(Kmers)){
    Kmer_counts[i]<-count_pattern(genome,Kmers[i])
  }
  #count of the most frequent kmers
  max_count <- max(Kmer_counts)
  #Get most frequent Kmers
  frequent_kmers <- Kmers[Kmer_counts == max_count]
  return(frequent_kmers)
}
my_genome <- rand_Genome(300)
find_frequent_kmers(my_genome,3)
```

The first step of the function is defining the Kmers. Since it is any variable amount of nucleotides, we can use the argument genome and k. While the line of code seems redunant, the kmers \<- generate_k\_mers(genome,k) helps us define the variable when we call it later. Then we nest this variable into Kmers (capital K) to assign it the command unique. This ensures each kmer is unique and not redundant.

With Kmers defined, its important to count the kmers. To accomplish this, we need to assign kmers to a new variable called Kmer_count. This variable will be assigned to the command rep. This will replicate the length of kmers and assign the count of 0.

After setting Kmers_count, I set a for loop that ran through the entire length of the variable. I then assigned the count pattern function (defined in the last challenge) to Kmers_count with genome and Kmers as the arguments. This will essentially run through all of kmers and assign a count to Kmer_counts. After running the for loop, I use max to sort out the kmers that were most frequent and assign it to a new variable called max_count. The last step is to create a new variable called frequent_kmers which will be returned at the end of the function. This variable will redefine kmers as any kmer_counts that were equal to the max count. This ensures that when we call the kmers, it will pull only the counts that were equal to the maximum (or most frequent) no matter how many kmers we define.

Once the function is created the only thing left to do is pull a random genome and call it in the function with a defined list of kmers. Kmers allows us to set the amount of nucleotides to be varied and we can finally set this in the final line of code.

## Challenge 2

The goal of this code block is to write a function that can input a string of nucleotides and output a reverse of it.

```{r}
reverse_complement <- function(genomeSubstring){
  #reverse genomeSubstring
  splits <- strsplit(genomeSubstring,"")[[1]]
  reversed_splits <- rev(splits)
  rev_complement <- c()
  #use if statements to construct compliment of reversed genomeSubstring
  for (i in 1:length(reversed_splits)){
    if (reversed_splits[i] == "A"){
      rev_complement <- append(rev_complement,"T")
    }
    if(reversed_splits[i] == "T"){
      rev_complement <- append(rev_complement,"A")
    }
    if(reversed_splits[i] == "C"){
      rev_complement <- append(rev_complement, "G")
    }
    if(reversed_splits[i] == "G"){
      rev_complement <- append(rev_complement, "C")
    }
  }
rev_complement <-  paste(rev_complement, collapse = "")
return(rev_complement)
}


reverse_complement("CGTTA")
```

To start, I created a function with genomeSubstring in parenthesis. Since this function is simply reversing a string and requires no numerical input, we only need that one argument. The first part of the function involves splitting the string. This helps us later on when we need to run a for loop that examines each part of the string individually. To do this we will create a new variable called splits and assign the genomeSubstring to it (with the strsplit function).

From there we then assign splits to a new variable called reversed_splits with the rev function. This will reverse the order of the string by element.

Now that the string is reversed, we need to find the complement pair. This will involve creating a for loop where we run through the entire length of reversed_splits. We then construct if statements that follow complimentary base pairing. So for example, if the reversed string had an adenine it would be complimented with a thymine (and a thymine would be complimented with an adenine). A guanine would be paired with a cytosine and vice versa. To construct this reverse complmenet, we need to create a new variable called rev complmenet and contain it in a list (denoted as c()). Collecting the strings, we then assign the append function to variable nested under each if statement. Once the for loop is finished, the rev complement is returned.

With the function complement, the only thing left to do is call the function and input a string of nucleotides as the genome string.

# Origin of Replication III

## Challenge 1

The goal of this challenge is to identify what parts of a genome represent the origin of replication. In the genome, these are representend as DNABoxes. These parts of the genome can be identified by clumps of frequent nucleotides in a string. In our last challenge, we created a function that found frequent k_mers (variable amount of nucleotides) in a genome. However this function will find all the potential k_mers in a subsetted window (defined by its length, L) and add it to a count in a dictionary. From there it will create a list of potential candidates based on the frequency of the k_mers.

Creating the k_mer dictionary requries only k as an argument. The variable nucleotide is defined and assigned to k_mers_dict. From there the variable makes sure that all the k_mers are unique. This ultimately creates a dictionary that includes all potential combinations of A,C,G,T (the type of nucleotides) ranging from 0 to 9.

```{r}
initialize_k_mer_dict <- function(k){
  nucleotides <- c("A", "C", "G", "T")

  k_mers_dict <- expand.grid(rep(list(nucleotides), k)) %>%
    unite("k_mers", everything(), remove = TRUE, sep = "") %>%
    unique() %>%
    mutate(count = 0)
  
  return(k_mers_dict)
}

k_mers_dict <- initialize_k_mer_dict(9)

k_mers_dict
```

With our dictionary created, we can now use it to find our clumps. Our function will use four arguments. The genome will represent whatever genome we import (in this challenge it is rand_Genome but it could be used on the e coli genome as well). L will represent the length of the window. The window is the amount of nucleotides in a window we will examine for frequent k_mers (candidates for our DNABoxes). K will represent the length of our k_mers. T is the threshold (the count or frequency of specific k_mers) we must reach before considering the k_mer as a candidate.

We will also describe the amount of characters (nchar) of the genome to a variable known as ng. The initial window is designed to be a string sub of the genome and starts with the first nucleotide. It extends to L which will be the length of our genome.

With the intial window defined, we need to develop a pattern that will detect the most frequent k_mers in that window and add it to the dictionary. To do this, we will develop a for loop that runs through every nucleotide from the first to L-(k-1). This will be hard to visualize but the idea is our for loop will run to the start of the very last K_mer. If our K_mer is nine for example, we will need to get the value of 8 to cutoff at the first nucleotide of the last k_mer. L-K would subtract nine while the -1 equals out to -8 which gets us to the first nucleotide.

With the window of our for_loop defined, we have to create a pattern that will look through an entire k_mer not just one nucleotide at a time. This will be defined as curr_pattern and is a str_sub of the initial window. It begins with the first nucleotide of the k_mer (i) and i + (k-1) which is the last nucleotide of the k_mer. For example if our k_mer is nine, we want the ninth nucleotide to be the last in our pattern. To get this we cant just to k because that implies we are going to the tenth element. To get to the ninth element, we have to add a k to nucleotide (gets us ten) and minus 1 to get to the end of the k_mer. Using our pattern in the initial window, we will call the function count_pattern and assign the k_mer to a variable called curr_count. Then we need to match our pattern (the k_mer we read in our window) with a table in our dictionary. This is accomplished by searching through our dictionary (my_dict) and equaling it to curr_pattern. We will save this matched k_mer to the variable curr_row. Finally we call the curr_Row and assign it to the count we found.

Once the for loop has gone through our entire window, we create a new variable called candidates. This will look through our dictionary for any k_mers that have a count that exceed our threshold (t) and then pull those k_mers. We are now slowly building a list of candidate k_mers that could be our origin of replication based on their frequency.

However, this only covers the initial window. We have to slide the window through our thereotical genome. To do this is a bit complex. We need to shift our window by 1 k_mer by removing the first nucleotide and starting at the second nucleotide. We then need to add

```{r}
clump_finding <- function(genome, L, k, t){
  #Generate a kmer dictionary
  my_dict <- initialize_k_mer_dict(k)
  ng <- nchar(genome)
  #Define the initial window of nucleotides in the clump
  initial_window <-  str_sub(genome, 1, L)
  #Get Kmer counts for initial window and fill in column of dictionary
  #str_sub(initial_window,i, i + k-1)
  #str_sub(initial_window, i, i + (k+1))
  #Our for loop should start with 1 (i) and end with L. 
  for (i in 1:(L-(k-1))){
    curr_pattern <- str_sub(initial_window, i, i + (k-1))
    curr_count <- count_pattern(initial_window, curr_pattern)
    curr_row <- which(my_dict$k_mers == curr_pattern)
    my_dict$count[curr_row] <- curr_count
  }
    #look for current pattern in kmer column of dictionary
 candidates <- my_dict%>%
    filter(count>=t)%>%
    pull(k_mers)
    #write current count in count column of that kmer row.
  for (j in 2:(ng - (L-2))){
    remove_kmer <- str_sub(genome,j-1, j - 1 + (k-1))
    add_kmer <- str_sub(genome,j + (L-2) - (k-1), j + (L-2))
    
    rem_row <- which(my_dict$k_mers == remove_kmer)
    add_row <- which(my_dict$k_mers == add_kmer)
    
    my_dict$count[rem_row] <- my_dict$count[rem_row] - 1
    my_dict$count[add_row] <- my_dict$count[add_row] + 1
    
    if(my_dict$count[add_row] >= t){
      candidates <- append(candidates, add_kmer)
      candidates <- unique(candidates)
    }
  }
  return(candidates)
}  

myGenome <- rand_Genome(700)
clump_finding(myGenome,500, 2, 9)

```

Since it should work for every genome you put in the function, avoid defining the genome right away. DNABoxes are usually divisible by three (due to presence of codons) but is not always nine (depends on species). String sub will allow us to grab a part of a string vs string split which actually splits the string.\

# Origin of Replication IV

## Challenge 1

```{r}
SKEW <- function(genome){
  ng <- nchar(genome)
  skew_value <- rep(0,ng)
  for (i in 2:(ng - 1)){
    genome_string <- str_sub(genome, i, i)
    if (genome_string == "G"){
      skew_value[i] = skew_value[i-1] + 1
    }
    if (genome_string == "C"){
      skew_value[i] = skew_value[i-1] - 1
    }
    else{
      skew_value[i] = skew_value[i-1] 
    }
  }
return(skew_value)
}

skew_genome <- rand_genome
SKEW(skew_genome)
```

```{r}
find_skew_min <- function(genome){
  ng <- nchar(genome)
  skew_value <- rep(0,ng)
  for (i in 2:(ng - 1)){
    genome_string <- str_sub(genome, i, i)
    if (genome_string == "G"){
      skew_value[i] = skew_value[i-1] + 1
    }
    if (genome_string == "C"){
      skew_value[i] = skew_value[i-1] - 1
    }
    else{
      skew_value[i] = skew_value[i-1] 
    }
  }

  min_skew <- min(skew_value)
  min_skew_loc <- which(skew_value == min_skew)
  return(min_skew_loc)  
}
skew_genome <- rand_genome
find_skew_min(skew_genome)
```

# Origin of Replication V

## Challenge 1

```{r}
getDistance <- function(stringA, stringB){
  k <- nchar(stringA)
  distance <- 0
   if(nchar(stringA) != nchar(stringB)){
        stop("ERROR sequences do not have the same length")
   }
 for (i in 1:k){
   if (str_sub(stringA,i,i) != str_sub(stringB,i,i)){
     distance <- distance + 1 }
 }
  return(distance)
}

getDistance("ACCCCC","AGCGGA")

```

## Challenge 2

```{r}
find_near_pattern <- function(genomeString, pattern, d){
  d = 0
  
}
```
